sdk/core/core/internal/az_retry_internal.h

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef _az_RETRY_INTERNAL_H
+#define _az_RETRY_INTERNAL_H
+
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+AZ_NODISCARD AZ_INLINE int32_t _az_retry_calc_delay(
+    int16_t attempt,
+    int32_t retry_delay_msec,
```

Comment:
Shouldn't this be in seconds? Why would anyone care about ms retry resolution from an embedded client?
Then we can move to int16_t (I don't think anyone would want to wait 18hrs before a retry.)

Here's one retry model I designed (it is in ms but doesn't need to):

x axis is the retry count; y is the delay for that retry. Total time waited is the integral: the key thing is to limit the total # of retries to never exceed 1, max 2 per second (which would quickly drain battery).

![image](https://user-images.githubusercontent.com/8635911/75931737-8454aa80-5e2a-11ea-9614-5f8c614f0587.png)

Reply:
Even more for my argument to move to seconds: `Retry-After` is also in seconds.

Reply:
@CIPop is your y axis representing seconds or msec?  Was your implementation well received/liked by clients?

Reply:
'retry-after-ms' and 'x-ms-retry-after-ms' are in milliseconds.
I also see the problem, but currently our PAL layer operates on int64_t msec. (az_platform_clock_msec() and az_platform_sleep_msec() return it)

We have a plan to review this aspect, to make SDK usable on both platforms with only seconds resolution, and 32-bit clock, and on desktop-like platforms, such as windows and linux, with millisecond resolution.

Ideally, int32 one_hundred_millisecond unit should be good enough for all applications. Then, if a platform only supports 0.5 seconds, or even seconds - sleep() would sleep 1 second if you ask it to sleep for `1 * one_hundred_millisecond`. Then, int32_max is ~6.8 years.

But all this requires deeper thinking, and we will most likely change it soon. But not in the scope of this PR; taking care of it tactically in one place is not going to solve the problem, but will bring inconsistency.

Reply:
I am consider this PR feedback as 'done' for the scope of this PR, and not going to do more commits,and I'm going to resolve this conversation. Please revert it or ping me if you disagree and think that I should not check this in.

Reply:
Not that big of a deal I think. I'd like to spend some time on an actual 8bit platform before pushing for optimizations such as the ones I proposed.
Didn't know about retry-after-ms - that doesn't seem to be a HTTP standard header. If some of the servers use those then I see good reason to hold ms resolution.

@gilbertw it's in msec. The implementation was transformed over several iterations to avoid certain issues detected by clients such as the max # of retries / second (high CPU/fast battery consumption). 
Obviously retry-after is the best way since the service has a lot more knowledge of its state (e.g. throttled - retry tomorrow). 

When the network is partitioned though the retry policy must be tuned by the device manufacturer as a function of number of devices, time to re-connect all devices, network latency, max timeouts within the network infrastructure, etc. The retry model above is favoring a busy service to massively reconnect a high number of devices (the randomization around +/-5 sec) with low TCP/TLS impact. This is quite common in IoT when a large area becomes partitioned then is reconnected.